### DISCUSSION TITLE
Better handling of dev scripts and dependencies

### DISCUSSION THREAD

**Discussion by gh-andre** [npm/feedback#266] [CLI] [answered]
A typical app will have a ton of development scripts and dependencies used for building, testing and debugging. Current `package.json` format does provide limited support for development dependencies in `devDependecies`, but when an app is built and published, there is no way to remove all these scripts and dependencies, which is not only confusing in production environments, but can also make the app more vulnerable if somebody runs a development script in production by mistake and it works.

Here's the default list of scripts generated by NestJS and it has just one production script - `node dist/main`. Everything else is development stuff:
```
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "nest build",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "start": "nest start",
    "start:dev": "nest start --watch",
    "start:debug": "nest start --debug --watch",
    "start:prod": "node dist/main",
    "lint": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:cov": "jest --coverage",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --config ./test/jest-e2e.json"
  },
```
At the very least there should be a reciprocal `devScripts` property to split development and production scripts, so build process can incorporate a simple step of stripping off `devScripts` and `devDependencies`.

A better approach would be to have a single `dev` property with the same property names as in the main part of `package.json`, but for development, as shown below, so they can be stripped off with a single filter, as opposed to having to hunt down all development script names to filter them out:
```
{
  "scripts": {...},
  "dependencies": {...},
  "dev":  {
    "scripts": {...},
    "dependencies": {...}
  }
}
```
Finally, even better approach would be if the npm CLI would have an option to strip off development stuff when packing the app, so a TypeScript app would just pack generated JavaScript from the build folders (e.g. `dist`, etc), along with all accompanying stuff, like docs, template views, static files, binaries, etc.

**Comment by ljharb** [comment]
Generally speaking, production doesn't necessarily even have npm - the production artifact is built with `npm install --production` (or `npm prune --production`), and deployed wholesale. Additionally, how would anyone run an npm script in a production environment when most users should not have production shell access in the first place?

**Reply by gh-andre** [reply]
On `npm install --production`, it requires somebody to be mindful of adding this flag, as opposed to have a package that is production-only.

On `npm prune --production`, it removes "extraneous packages" blindly - just because they are not referenced. Again, having explicit references just for production is a more robust option.

>  how would anyone run an npm script in a production environment when most users should not have production shell access in the first place?

Not sure I'm following. Large apps are installed and upgraded by DevOps/NOC folks all the time, either directly or by maintaining scripts that do this on their behalf, which is not different from people installing these apps - it's just automation. Those folks and their automation are not some random users, but dedicated people who are meant to do maintenance and the less options they have to remember, the more robust will be the installation.

**Reply by ljharb** [reply]
Right, but those scripts aren't generally part of the project being deployed - they're kept separate.

**Reply by gh-andre** [reply]
Yes, they are indeed, but that's where the discrepancy lies - these DevOps scripts handle tasks that are used for publishing apps - something that `npm` is supposed to handle. Let me elaborate once more.

TypeScript and other advancements dramatically changed the lifecycle of Node apps - now they have many steps that didn't exist before - build, debug, test, etc. None of them have anything to do with the app's functionality and have no place in the package, as you can see in the OP example.

`npm` recognizes this half-way - hence `devDependencies`, as are some of those `--production` flags, but those who want to package only the product of the app's build process have to come up with separate DevOps scripts that strip off development `scripts` entries and `devDependencies` before they pack their app. The fact that these scripts have to parse and modify `package.json` is an indicator that there's functionality missing in the package manager for this.

Having a built-in process that will publish apps only with the stuff that packages need to install, verify installation, run and uninstall, will eliminate many of those home-grown DevOps scripts.

**Reply by ljharb** [reply]
I'm missing that part - why does it matter if scripts or devDeps are present, since consumers are (without exception) expected (i'd even say, mandated) to not run or install those?

**Reply by gh-andre** [reply]
Well, besides just having a package manager being logically complete for intended uses, have you tried to convince PA DSS auditors that having build/test stuff in production is fine because users are strongly instructed not use it?

**Reply by ljharb** [reply]
I have no idea what "PA DSS auditors" are, but the simple answer to me seems to be "the `npm` binary is not available in production, so those fields are irrelevant"

**Reply by gh-andre** [reply]
> I have no idea what "PA DSS auditors" are, but the simple answer to me seems to be "the `npm` binary is not available in production, so those fields are irrelevant"

Ok. Thanks for your thoughts.

**Comment by MylesBorins** [comment, marked as answer]
hey @gh-andre

As mentioned in the other thread when installing you can run `npm install --production` to ensure that you dev dependencies are not installed. As of today there are no commands or processes that would augment the `package.json` at any point aside from `npm install` and `npm uninstall`, more specifically there is no concept of pre / post production iterations of the package.json.

It seems like you have two distinct proposals

* devScripts
* `dev` section for scripts + deps

Both of these options would be non-backwards compatible, but perhaps that doesn't matter for applications vs packages. I'll be honest that I am not extremely sold on these two concepts BUT I'm not the person who makes these decisions 😇. If you want to pursue these as new features in the npm CLI you can open an issue proposing them at https://github.com/npm/rfcs. For an RFC it helps to be very specific, thankfully there is a template you can follow.

It seems to me like this is something like could be handled in userland with a pre-processing script modifying the package.json when preparing the application for prod.

Here is a quick example that goes through and deletes all devDeps as well as all scripts that include the keyword `test` or `dev`

```js
const {writeFileSync} = require('fs');
const pkg = require('./package.json');
delete pkg.devDependencies;
Object.keys(pkg.scripts).forEach(script => {
  if (script === 'test' || script.includes('dev')) delete pkg.scripts[script];
})
writeFileSync('./package.json', JSON.stringify(pkg, null, 2))
```

**Reply by gh-andre** [reply]
@MylesBorins  Thanks for the insights.

> It seems like you have two distinct proposals

Just two different possible solutions for the same problem. One follows the pattern of `devDependencies` and one more cleanly separating _all_ dev stuff from the rest of the package contents.

> Both of these options would be non-backwards compatible

They are, actually. Not using either of those options just leaves one where they are today and if there was an `npm` option to publish without dev stuff, they would publish just as they are doing it today. Those who wish to move to a cleaner separation of the dev and production stuff would produce cleaner published packages without defunct dev references in `scripts`.

> It seems to me like this is something like could be handled in userland

A lot of things can be done with `node_modules` and `package.json` just by reaching into internal folders or editing files that just happen to be there. It doesn't mean that it should be done this way, though, simply because when this structure changes, and it most certainly will at some point, all those scripts will go bad. Doing it via the package manager is the preferred way, hence this and the other thread.

> Here is a quick example

Yes, editing `package.json` is the only way to deal with this now, but I will mention for anyone who comes across this thread that you most certainly don't want to do string pattern matching because words like `test` and `dev` will occur in most unexpected places. Listing production scripts in an array of complete strings and removing all others from the existing `scripts` property will less likely backfire by removing something that is needed.
